### PR TITLE
posix: implement fcntl(..., F_SETFL, ...)

### PR DESCRIPTION
**Pull request by marcinslusarz**
... and add required locking in F_GETFL.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/406)

### CODE REVIEW

File: src/libpmemfile-posix/fcntl.c [hunk 1]
```diff
@@ -102,8 +107,61 @@ pmemfile_fcntl(PMEMfilepool *pfp, PMEMfile *file, int cmd, ...)
...
+				errno = EINVAL;
+				return -1;
+
+			}
+#undef SUPPORTED_FLAGS
+
+			os_mutex_lock(&file->mutex);
+
+			if (fl_flags & PMEMFILE_O_APPEND)
+				file->flags |= PFILE_APPEND;
```
Comment (GBuella) [inline src/libpmemfile-posix/fcntl.c]: This changes are shared among dup'ed fds.
So, to be clear, these are the *file status flags*, shared by dup'ed fd's, and the *file descriptor flags* are not shared, and those should be stored in the libpmemfile layer. Right?
And the only *file descriptor flag* we know about is the close-on-exec flag, if I interpret things right.

So, `F_SETFD` should be handled in libpmemfile, and not passed on to libpmemfile-posix, I guess.

File: src/libpmemfile-posix/fcntl.c [hunk 2]
```diff
@@ -102,8 +107,61 @@ pmemfile_fcntl(PMEMfilepool *pfp, PMEMfile *file, int cmd, ...)
...
+					PMEMFILE_O_TRUNC | PMEMFILE_O_SYNC |
+					PMEMFILE_O_DSYNC);
+
+			if (fl_flags & PMEMFILE_O_ASYNC) {
+				ERR("setting O_ASYNC is not supported");
+				errno = EINVAL;
+				return -1;
+			}
+
+#define SUPPORTED_FLAGS (PMEMFILE_O_APPEND | PMEMFILE_O_DIRECT | \
```
Comment (GBuella) [inline src/libpmemfile-posix/fcntl.c]: How about `static const int supported_flags`?

File: src/libpmemfile-posix/fcntl.c (hunk 1 above)
Comment (marcinslusarz) [reply src/libpmemfile-posix/fcntl.c]: You are right, but it's not relevant to this patch :). This patch is about F_SETFL, not F_SETFD.

Comment (GBuella) [reply src/libpmemfile-posix/fcntl.c]: When I looked at it, I just wasn't sure if these changes should be OK with dup, or not, but if you agree too, then OK.

File: src/libpmemfile-posix/fcntl.c (hunk 2 above)
Comment (marcinslusarz) [reply src/libpmemfile-posix/fcntl.c]: Sure. Done.

**Comment by sarahjelinek** [comment]
:lgtm:

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/406#-:-KvcAx4DA31ajxpw-rae:bnfp4nl)*

**Pull request merged**